llvm/lib/Target/DirectX/DXILIntrinsicExpansion.cpp

```diff
@@ -70,31 +71,57 @@ static bool expandAbs(CallInst *Orig) {
   return true;
 }
 
-static bool expandIntegerDot(CallInst *Orig, Intrinsic::ID DotIntrinsic) {
-  assert(DotIntrinsic == Intrinsic::dx_sdot ||
-         DotIntrinsic == Intrinsic::dx_udot);
-  Intrinsic::ID MadIntrinsic = DotIntrinsic == Intrinsic::dx_sdot
-                                   ? Intrinsic::dx_imad
-                                   : Intrinsic::dx_umad;
+static bool expandDotIntrinsic(CallInst *Orig, Intrinsic::ID DotIntrinsic) {
```

Comment:
for simplicity I would keep a seperate  `expandIntegerDot` and `expandFloatDot`. Doing it this way is a little weird considering you already have a conditional to seperate behavior via the switch cases.  Then you merge back only to seperate out the behaviors again via contional `if (EltTy->isIntegerTy()) {`